### DISCUSSION TITLE
Network errors with the offlineExchange

### DISCUSSION THREAD

**Discussion by bigescalade** [urql-graphql/urql#1520] [Q&A] [answered]
Hello there. We are using the offlineExchange to bring some offline capabilities in our app. The docs state that we no longer get Network Errors when using the offlineExchange

`When a query fails with a Network Error, which indicates that the client is offline — either because navigator.onLine is false or because the error message indicates an offline error — the offlineExchange won't deliver the error for this query to avoid it from being surfaced to the user.`

is there any way to get the Network Errors while using the offlineExchange? I tried using the errorExchange but failed network requests don't seem to get there.

We want to implement an error screen if the request fails for any reason (offline, timed out due to slow connection etc) with a retry button. We can determine if the device is offline but without Network Errors we are having trouble handling slow connection/timeouts

**Comment by kitten** [comment, marked as answer]
Are you placing the `errorExchange` before or after the `offlineExchange`? If it's not showing you the network errors then it's likely before the `offlineExchange`; as operations travel forward in the list of exchanges and results backwards by the time the network error reaches the `offlineExchange` it may be swallowed if you're looking at failed queries, so you'll have to place it after the `offlineExchange`.

We're also still thinking about how to improve this process. Our original idea behind this mechanism of swallowed errors is that the UI should be completely resilient to offline states. However, we realise now that sometimes there need to be escape hatches when an incomplete result (partial or empty; no matter) just isn't enough.

Until then the `errorExchange` should indeed help you spot which queries have failed. Furthermore, if you examine the `offlineExchange`'a code, which operates on top of Graphcache's normal `cacheExchange` it may help you identify what's going on: https://github.com/FormidableLabs/urql/blob/6e6ac18d96347aedac32162d6d8096050069ef6a/exchanges/graphcache/src/offlineExchange.ts#L166

Specifically, it filters out failed mutation results before the `cacheExchange` can see them. And it filters out offline query errors before they can reach the rest of the exchanges (or rather, the app)

While the mechanism for mutations is pretty solid in my opinion, since optimistic mutations can safely remain optimistically applied indefinitely until the network connection is restored, it's offline queries that are not at their full potential yet, as long as we swallow the errors.

**Reply by bigescalade** [reply]
thanks for your very detailed reply. you are right i had the errorExchange before, putting it after I can now access all the errors. I can make this work until the escape hatch is implemented 😄
